### PR TITLE
Add validation for Decimal256

### LINKED ISSUE #2112: Do value validation for Decimal256
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently we skip value validation for precision > 38 for Decimal256. We should do it.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2112.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/datatypes/datatype.rs [hunk 1]
```diff
@@ -298,6 +299,49 @@ pub const MAX_DECIMAL_FOR_EACH_PRECISION: [i128; 38] = [
     99999999999999999999999999999999999999,
 ];
 
+/// `MAX_DECIMAL_FOR_LARGER_PRECISION[p]` holds the maximum integer value
+/// that can be stored in [DataType::Decimal256] value of precision `p` > 38
```
Comment (viirya) [inline arrow/src/datatypes/datatype.rs:303]: Depends on #2093 which adds `DataType::Decimal256`

Comment (viirya) [reply arrow/src/datatypes/datatype.rs:303]: So this causes doc failure in CI.

Comment (alamb) [reply arrow/src/datatypes/datatype.rs:305]: I took the liberty of merging this PR up from master.

File: arrow/src/array/data.rs [hunk 2]
```diff
@@ -1018,6 +1018,7 @@ impl ArrayData {
                 }
                 Ok(())
             }
+            // TODO: call validate_decimal256_precision for Decimal256 type
```
Comment (alamb) [inline arrow/src/array/data.rs:1021]: Is this still a TODO for this PR?

**Review by alamb — APPROVED** [review]

File: arrow/src/array/data.rs (hunk 2 above)
Comment (viirya) [reply arrow/src/array/data.rs:1021]: I was waiting for #2093 where added `Decimal256` type, so I can call `validate_decimal256_precision` here. Let me update this.

File: arrow/src/array/array_decimal.rs [hunk 3]
```diff
@@ -352,7 +352,7 @@ impl From<ArrayData> for Decimal256Array {
         assert_eq!(
             data.buffers().len(),
             1,
-            "Decimal128Array data should contain 1 buffer only (values)"
+            "Decimal256Array data should contain 1 buffer only (values)"
```
Comment (viirya) [inline arrow/src/array/array_decimal.rs:355]: fix typo

File: integration-testing/src/lib.rs [hunk 4]
```diff
@@ -640,10 +640,10 @@ fn array_from_json(
                         let decimal =
                             Decimal256::try_new_from_bytes(*precision, *scale, &bytes)
                                 .unwrap();
-                        b.append_value(&decimal)
+                        b.append_value(&decimal)?
                     }
                     _ => b.append_null(),
-                }?;
+                };
```
Comment (viirya) [inline integration-testing/src/lib.rs:646]: Not sure why this was not caught in CI previously.

Comment (alamb) [reply integration-testing/src/lib.rs:646]: I think it was a logical conflict: https://github.com/apache/arrow-rs/pull/2121

**Comment by viirya** [comment]
Thanks @alamb

**Pull request merged**